Implement booking and update users models

ihatelactose opened the pull request:
### Description

- Implemented the bookings model
- Created the bookingsDao for db interaction
- Updated the users model to form associations with bookings model
- Wrote migrations for bookings and altering bookings

### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added]
- [ ] Relevant documentation is changed or added (and PR referenced)

## curls
### cURL for Create Booking

- `toLocation` is sent as a stringified array, which then is parsed and used for `toLocation` lat and lng
- `fromLocation` is sent as a stringified array, which then is parsed and used for `fromLocation` lat and lng

curl
```shell
curl --location --request POST 'http://localhost:9000/booking' \
--header 'Content-Type: application/json' \
--data-raw '{
    "toLocation": "[-76.984722, 39.807222]",
    "fromLocation": "[-76.984722, 39.807222]",
    "userId": 2
}'
```

Response
```json
{
    "status": "Booking created successfully!"
}
```

DB update
<img width="1129" alt="Screenshot 2022-05-10 at 6 05 40 PM" src="https://user-images.githubusercontent.com/104886701/167629680-b29e674c-96a5-4e06-8015-604ebc2c5991.png">

### cURL for Get Booking
curl
```shell
curl --location --request GET 'http://localhost:9000/booking/1'
```

Response
```json
{
    "booking": {
        "id": 1,
        "fromLocation": {
            "type": "Point",
            "coordinates": [
                39.807222,
                -76.984722
            ]
        },
        "toLocation": {
            "type": "Point",
            "coordinates": [
                73.90242886592246,
                18.479759440702946
            ]
        },
        "userId": 2
    }
}
```

### cURL for Get all Booking
curl
```shell
curl --location --request GET 'http://localhost:9000/users/2/bookings'
```

Response
```json
{
    "bookings": {
        "all_bookings": [
            {
                "id": 1,
                "fromLocation": {
                    "type": "Point",
                    "coordinates": [
                        39.807222,
                        -76.984722
                    ]
                },
                "toLocation": {
                    "type": "Point",
                    "coordinates": [
                        73.90242886592246,
                        18.479759440702946
                    ]
                },
                "userId": 2
            }
        ],
        "total_count": 1
    }
}
```

### cURL for Create User
curl
```shell
curl --location --request POST 'http://localhost:9000/users/register' \
--header 'Content-Type: application/json' \
--data-raw '{
    "id": 3,
    "firstName": "Test",
    "lastName": "User",
    "email": "test@user.com"
}'
```

### cURL for Get Nearby Cabs
curl
```shell
curl --location --request POST 'http://localhost:9000/cab/nearby' \
--header 'Content-Type: application/json' \
--data-raw '{
    "current_location": "[18.479759440702946, 73.90242886592246]",
    "radius": 100
}'
```

response
```json
{
    "cabs_nearby": [
        {
            "first_name": "John",
            "distance_m": "1943"
        },
        {
            "first_name": "Jane",
            "distance_m": "1521"
        },
        {
            "first_name": "Clark",
            "distance_m": "3527"
        }
    ]
}
```

Commit:
feat: Implemented a basic implementation of Booking;
- Created Booking Model, SQL, and relevant migration
- Created BookingDAO for interaction with the database
- Created route for Booking interactions
- Modified users model for supporting relations/associations with Bookings

Commit:
feat: Implemented creating user and updated userDao

Commit:
refactor: updated camel case to snake case in SQL files and updated bookings and users model with the same

Commit:
fix: updated bookings model to have GEOMETRY types instead of STRING

Commit:
fix: updated the booking route handler to use raw query and manually insert Point data into the table

Commit:
update: fixed the typo in booking/routes.js

ihatelactose commented:
### cURL for Booking

- `toLocation` is sent as a stringified array, which then is parsed and used for `toLocation` lat and lng
- `fromLocation` is sent as a stringified array, which then is parsed and used for `fromLocation` lat and lng

curl
```shell
curl --location --request POST 'http://localhost:9000/booking' \
--header 'Content-Type: application/json' \
--data-raw '{
    "toLocation": "[-76.984722, 39.807222]",
    "fromLocation": "[-76.984722, 39.807222]",
    "userId": 2
}'
```

Response
```json
{
    "status": "Booking created successfully!"
}
```

DB update
<img width="1129" alt="Screenshot 2022-05-10 at 6 05 40 PM" src="https://user-images.githubusercontent.com/104886701/167629680-b29e674c-96a5-4e06-8015-604ebc2c5991.png">

Commit:
refactor: moved the insert query for booking from route handler to bookingDao

tousifws commented on lib/daos/bookingDao.js:
```diff
@@ -0,0 +1,77 @@
+ */
+export const findAllBookings = async (page, limit) => {
+    const where = {};
+    const totalCount = await bookings.count({ where });
+    const allBookings = await bookings.findAll({
+        attributes,
+        where,
+        offset: (page - 1) * limit,
+        limit
+    });
+
```
Sequelize has `findAndCountAll`  [https://sequelize.org/docs/v6/core-concepts/model-querying-finders/#findandcountall](https://sequelize.org/docs/v6/core-concepts/model-querying-finders/#findandcountall) Let's use that one

tousifws commented on lib/daos/bookingDao.js:
```diff
@@ -0,0 +1,77 @@
+
+    await sequelize.query(
+        `
+        INSERT INTO bookings VALUES
+            (
+                DEFAULT,
+                ST_SRID(ST_GeomFromText('POINT(:fromLat :fromLng)'), 4326),
+                ST_SRID(ST_GeomFromText('POINT(:toLat :toLng)'), 4326),
+                :userId,
+                DEFAULT
+            )
+        `,
```
Use sequelize model to do the insertion

tousifws commented on lib/routes/users/routes.js:
```diff
@@ -1,10 +1,50 @@
+            auth: false,
+            validate: {
+                payload: Joi.object({
+                    firstName: stringSchema,
+                    lastName: stringSchema,
+                    email: emailSchema,
+                    id: numberSchema,
+                    oauth_client_id: numberSchema
```
We accept payloads in snake case. Update them everywhere

tousifws commented on resources/v2/02_alter_bookings.sql:
```diff
@@ -0,0 +1 @@
+  ALTER TABLE bookings MODIFY id INT UNSIGNED NOT NULL AUTO_INCREMENT, AUTO_INCREMENT=1;
```
What is this for? By default, the starting value for `AUTO_INCREMENT` is 1 https://www.w3schools.com/sql/sql_autoincrement.asp

tousifws commented on lib/daos/userDao.js:
```diff
@@ -8,6 +8,24 @@ const attributes = [
+ */
+export const createOneUser = async data => {
+    const { firstName, lastName, email, id, oauth_client_id } = data;
+    const newUser = await users.create({
```
Since the `newUser` is variable is being used, you can return the `users.create` without storing it in a variable

tousifws commented on resources/v2/01_bookings.sql:
```diff
@@ -0,0 +1,9 @@
+    id              INT UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
+    from_location   POINT NOT NULL SRID 4326,
+    to_location     POINT NOT NULL SRID 4326,
+    by_user         INT NOT NULL,
```
Let's use user\_id instead of by\_user

Commit:
feat: Created Cab and Drivers model with their respective migrations and sql files
- Created Cab model, migration, and sql query
- Created Driver model, migration, and sql query
- Changed the count and findAll to countAndFindAll in bookingDao.js
- Updated raw insert query to model query in bookingDao.js
- Updated route verification from camelCase to snake_case in overall routes
- Removed the 02_alter_booking.sql (was not doing anything since AUTO_INCREMENT is 0 by default)
- Updated the createOneUser Dao function to return immediately
- The bookings table column by_user is now user_id, updated the respective model and sql

tousifws commented on lib/routes/users/routes.js:
```diff
@@ -1,10 +1,50 @@
+            } catch (error) {
+                console.error(error);
+                return badImplementation();
+            }
```
You won't have to put in a try-catch block server error gets handled automatically. Update them all other places

tousifws commented on resources/v3/02_cabs.sql:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE cabs 
+ (
+   id               INT UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
+   driver_id_fk     INT UNSIGNED UNIQUE,
```
Field name should be just `driver_id` not `driver_id_fk`. And name the foreign key

Commit:
feat: Created updateCab, query for nearby cabs
- updated migrations from POINT to GEOMETRY
- updated cabDao, driverDao with find and update
- added more routes to /booking, /cab, and /driver

Commit:
feat: implemented finding cabs by cabType

tousifws commented on lib/daos/bookingDao.js:
```diff
@@ -0,0 +1,76 @@
+            'ST_SRID',
+            sequelize.fn(
+                'ST_GeomFromText',
+                `POINT(${parsedFromLocation[0]} ${parsedFromLocation[1]})`
```
Are we setting coordinates in the correct order [https://sequelize.org/api/v6/class/src/data-types.js~geometry](https://sequelize.org/api/v6/class/src/data-types.js~geometry)

GeoJson format: \[Longitude, Latitude\]

tousifws commented on lib/models/users.js:
```diff
@@ -33,4 +33,12 @@ module.exports = function (sequelize, DataTypes) {
+
+    User.associate = function (models) {
+        User.hasMany(models.bookings, {
+            foreignKey: 'by_user'
```
Isn't it `user_id`?

tousifws commented on utils/convertToPoint.js:
```diff
@@ -0,0 +1,4 @@
+export const convertToPoint = serializedPoint => ({
```
Is it being used anywhere?

tousifws reviewed (commented):
Please add the request response of all api end points

Commit:
refactor: updated the routes to NOT use trycatch, removed stale files, and updated the cabs.sql file

tousifws commented on lib/routes/cab/routes.js:
```diff
@@ -0,0 +1,149 @@
+                }
+            );
+
+            return h.response({ cabsNearby: nearby[0] });
```
Why return one car? We need all nearby cars right?

ihatelactose replied:
Yeah as mentioned that the response I get from the raw query is something like this
```json
{
    "nearby": [ "All the cabs" ],
    "nearby": [ "All the cabs" ]
}
```
Therefore to send just once I have done `nearby[0]`

tousifws commented on lib/routes/cab/routes.js:
```diff
@@ -0,0 +1,149 @@
+                }
+            );
+
+            return h.response({ cabsNearby: nearby[0] });
```
Why return one car? We need all nearby cars right?

tousifws commented on lib/routes/users/routes.js:
```diff
@@ -1,10 +1,64 @@
+            }
+        },
+        handler: async (request, h) => {
+            const { first_name, last_name, email, id, oauth_client_id } =
```
Aren't we getting them in camelCase?

tousifws replied:
Aren't we getting them in camelCase?

Commit:
refactor: updated the daos, routes, wrote tests, and updated the drivers.sql to include the status of driver

Commit:
refactor: removed the trycatch block from driver/routes

Commit:
fix: fixed the typos in the driver/test/routes.test.js file

tousifws commented on lib/routes/booking/routes.js:
```diff
@@ -0,0 +1,76 @@
+        method: 'GET',
+        path: '/',
+        options: {
+            description: 'Find all bookings',
+            notes: 'Find all booking',
+            tags: ['api', 'booking'],
+            auth: false
+        },
+        handler: async (request, h) => {
+            const bookings = await findAllBookings();
```
Add pagination. We use this plugin https://github.com/fknop/hapi-pagination

tousifws commented on seeders/06_bookings.js:
```diff
@@ -0,0 +1,71 @@
+                    'ST_SRID',
+                    Sequelize.fn(
+                        'ST_GeomFromText',
+                        'POINT(73.90242886592246 18.47975944070295)'
```
Add different nearby locations for every entry

tousifws commented on seeders/07_drivers.js:
```diff
@@ -0,0 +1,47 @@
+                    'ST_SRID',
+                    Sequelize.fn(
+                        'ST_GeomFromText',
+                        'POINT(73.90242886592246 18.47975944070295)'
```
Same location

tousifws commented on lib/routes/users/routes.js:
```diff
@@ -1,10 +1,64 @@
+            tags: ['api', 'users'],
+            auth: false
+        },
+        handler: async (request, h) => {
```
Add pagination

tousifws commented on resources/v3/01_drivers.sql:
```diff
@@ -0,0 +1,8 @@
+   first_name        VARCHAR(32) NOT NULL,
+   last_name        VARCHAR(32) NOT NULL,
+   current_location GEOMETRY NOT NULL,
+   driver_status    ENUM('open', 'transit', 'not_accepting') DEFAULT 'open'
```
just `status` not `driver_status`

Commit:
fix: updated the sql and created migrations, implemented paginations on certain routes

Commit:
tests: wrote tests for daos, routes, and updated utilities

Commit:
tests: updated the tests and fixed the methods for cache

tousifws commented on lib/routes/cab/tests/routes.test.js:
```diff
@@ -62,32 +74,21 @@ describe('/cab route tests', () => {
-        expect(newCab.driverId).toEqual(1);
-        expect(newCab.cabType).toEqual(cabType);
-        expect(newCab.cabNumber).toEqual(cabNumber);
+        expect(errors[0].message).toEqual('This driver does not exist.');
```
check for status code 404 also

tousifws reviewed (approved):
Resolve the comment then let's get this merged in

Commit:
fix: updated the seeds, fixed the distance received from scientific units to meters
